bin/preprocess.ml

```diff
@@ -0,0 +1,85 @@
+open Stdune
+open Import
+
+let doc = "Print the preprocessed output"
+
+let man =
+  [ `S "DESCRIPTION"
+  ; `P
+      {|$(b,dune preprocess FILE) build the given FILE and print the preprocessed output|}
+  ; `Blocks Common.help_secs
+  ]
+
+let info = Term.info "preprocess" ~doc ~man
+
+let change_ext f path =
+  let base, ext = Path.of_string path |> Path.split_extension in
+  let suffix = f ext in
+  Path.extend_basename base ~suffix |> Path.to_string
+
+let pp_with_ocamlc sctx project pp_ml =
+  let open Dune_engine in
+  let ocamlc = (Super_context.context sctx).ocamlc in
+  let env = Super_context.context_env sctx in
+  let argv = ["-stop-after"; "parsing"; "-dsource"; pp_ml; "-dump-into-file"] in
+  let open Fiber.O in
+  let* _ = Process.run ~env Process.Strict ocamlc argv in
+  let dump_file =
+    pp_ml |> change_ext (fun ext ->
+      let dialect =
+        Dialect.DB.find_by_extension (Dune_project.dialects project) ext
+      in
+      match dialect with
+      | None -> User_error.raise [ Pp.textf "unsupported extension: %s" ext ]
+      | Some (_, kind) ->
+        let open Import.Ml_kind in
+        match kind with
+        | Intf -> ".cmi.dump"
+        | Impl -> ".cmo.dump"
+    )
+  in
+  let dump_file_path = Path.of_string dump_file in
+  if not (Path.exists dump_file_path && Path.is_file dump_file_path) then
+    User_error.raise [ Pp.textf "cannot find a dump file: %s" dump_file ]
+  else
+    In_channel.with_open_text dump_file (fun ic -> Io.copy_channels ic stdout);
+    Sys.remove dump_file;
```

Comment:
You can use our `Temp` module to avoid having to remove stuff manually.

Reply:
I think we can't use `Temp` for this because we don't have a control over the filename.

It seems `try_paths` generates a filename with a random number, but ocamlc doesn't have an option to specify the name of the generated dump file.